caddyhttp/proxy/reverseproxy.go

```diff
@@ -260,8 +260,23 @@ func (rp *ReverseProxy) ServeHTTP(rw http.ResponseWriter, outreq *http.Request,
 		}
 		defer backendConn.Close()
 
-		go pooledIoCopy(backendConn, conn) // write tcp stream to backend
-		pooledIoCopy(conn, backendConn)    // read tcp stream from backend
+		// Proxy backend -> frontend.
+		go pooledIoCopy(conn, backendConn)
+
+		// Proxy frontend -> backend.
+		//
+		// NOTE: Hijack() sometimes returns buffered up bytes in brw which
+		// would be lost if we didn't read them out manually below.
+		if brw != nil {
+			if n := brw.Reader.Buffered(); n > 0 {
+				rbuf, err := brw.Reader.Peek(n)
+				if err != nil {
+					return err
+				}
+				backendConn.Write(rbuf)
```

Comment:
I guess you could consider it a premature optimization, because it only prevents executing whatever is done inside `Peek()` and `Write()`.
On the other hand doing nothing in case nothing needs to be done is something I've learned to be a very valuable optimization. Reason being that it's a) usually very straightforward to implement & understand and b) it automatically prevents regressions in the future if the internal implementations you rely on ever change.